src/lefout/lefout.cpp

```diff
@@ -811,6 +814,172 @@ void lefout::writeMTerm(dbMTerm* mterm)
   fprintf(_out, "    END %s\n", name.c_str());
 }
 
+void lefout::writePropertyDefinition(dbProperty* prop)
+{
+  std::string propName = prop->getName();
+  dbObjectType owner_type = prop->getPropOwner()->getObjectType();
+  dbProperty::Type prop_type = prop->getType();
+  std::string objectType,propType,value;
+  switch (owner_type)
+  {
+    case dbTechLayerObj:
+      objectType = "LAYER";
+      break;
+    case dbLibObj:
+      objectType = "LIBRARY";
+      break;
+    case dbMasterObj:
+      objectType = "MACRO";
+      break;
+    case dbMPinObj:
+      objectType = "PIN";
+      break;
+    case dbTechViaObj:
+      objectType = "VIA";
+      break;
+    case dbTechViaRuleObj:
+      objectType = "VIARULE";
+      break;
+    case dbTechNonDefaultRuleObj:
+      objectType = "NONDEFAULTRULE";
+      break;
+    default:
+      return;
+  }
+
+  switch (prop_type)
+  {
+  case dbProperty::INT_PROP:
+  case dbProperty::BOOL_PROP:
+    propType = "INTEGER";
+    break;
+  case dbProperty::DOUBLE_PROP:
+    propType = "REAL";
+    break;
+  case dbProperty::STRING_PROP:
+    propType = "STRING";
+    break;
+  default:
+    return;
+  }
+  fprintf(_out, "    %s %s %s ", objectType.c_str(),propName.c_str(),propType.c_str());
+  if(owner_type==dbLibObj){
+    fprintf(_out,"\n        ");
+    prop->writePropValue(prop,_out);
+    fprintf(_out,"\n    ");
+  }
+  
+  fprintf(_out, ";\n");
+}
+
+inline void lefout::writeObjectPropertyDefinitions(dbObject* obj,std::unordered_map<std::string,short>& propertiesMap)
+{
+  int bitNumber;
+  switch (obj->getObjectType())
+  {
+  case dbTechLayerObj:
+    bitNumber = 0;
+    break;
+  case dbLibObj:
+    bitNumber = 1;
+    break;
+  case dbMasterObj:
+    bitNumber = 2;
+    break;
+  case dbMPinObj:
+    bitNumber = 3;
+    break;
+  case dbTechViaObj:
+    bitNumber = 4;
+    break;
+  case dbTechViaRuleObj:
+    bitNumber = 5;
+    break;
+  case dbTechNonDefaultRuleObj:
+    bitNumber = 6;
+    break;
+  default:
+    return;
+  }
+  dbSet<dbProperty> properties = dbProperty::getProperties(obj); 
+  dbSet<dbProperty>::iterator pitr;
+  for (pitr = properties.begin(); pitr != properties.end(); ++pitr)
+  {
+    dbProperty* prop = *pitr;
+    if(propertiesMap.find(prop->getName())==propertiesMap.end())
```

Comment:
operator[] should default initialize the value so I don't think this is necessary.

Reply:
Done